src/opers.c

```diff
@@ -1821,2769 +1786,615 @@ static inline Obj CacheOper (
-    type1 = TYPE_OBJ_FEO( arg1 );
-    id1 = ID_TYPE( type1 );
-    type2 = TYPE_OBJ_FEO( arg2 );
-    id2 = ID_TYPE( type2 );
-
-    /* try to find an applicable method in the cache                       */
-    prec = INTOBJ_INT(-1);
-
-    do {
-      /* The next line depends on the implementation of INTOBJS */
-      prec = (Obj)(((Int)prec) +4);
-      method = 0;
-
-      /* recalculate cache each pass, in case of GC     */
-      cache = 1+ADDR_OBJ( CacheOper( oper, 2 ) );
-
-      /* Up to CACHE_SIZE methods might be in the cache */
-      if (prec < INTOBJ_INT(CACHE_SIZE))
-        {
-          for (i = 0;  i < 4*CACHE_SIZE; i+= 4) {
-            if (  cache[i+1] == prec && cache[i+2] == id1
-                  && cache[i+3] == id2 ) {
-              method = cache[i];
-#ifdef COUNT_OPERS
-              OperationHit++;
-#endif
-              break;
-            }
-          }
-        }
-      
-      /* otherwise try to find one in the list of methods                    */
-      if (!method)
-        {
-          if (prec == INTOBJ_INT(0))
-            method = CALL_3ARGS( Method2Args, oper, type1, type2 );
-          else
-            method = CALL_4ARGS( NextMethod2Args, oper, prec, type1, type2 );
-          
-          /* If there was no method found, then pass the information needed for
-             the error reporting. This function rarely returns */
-          if (method == Fail)
-          {
-            Obj args[2];
-            args[0] = arg1;
-            args[1] = arg2;
-            while (method == Fail)
-              method = CallHandleMethodNotFound( oper, 2, (Obj *) args, 0, 0, prec);
-          }
-          
-          /* update the cache */
-          if (method && prec < INTOBJ_INT(CACHE_SIZE))
-            {
-              Bag cacheBag = GET_METHOD_CACHE( oper, 2 );
-              cache = 1+ADDR_OBJ( cacheBag );
-              cache[4*STATE(CacheIndex)] = method;
-              cache[4*STATE(CacheIndex)+1] = prec;
-              cache[4*STATE(CacheIndex)+2] = id1;
-              cache[4*STATE(CacheIndex)+3] = id2;
-              STATE(CacheIndex) = (STATE(CacheIndex) + 1) % CACHE_SIZE;
-              CHANGED_BAG( cacheBag );
-            }
-#ifdef COUNT_OPERS
-          OperationMiss++;
-#endif
+    return method;
+}
+
+static inline void
+CacheMethod(Obj oper, UInt n, Obj prec, Obj * ids, Obj method)
+{
+
+    if (prec >= INTOBJ_INT(CACHE_SIZE))
+        return;
+    Bag   cacheBag = GET_METHOD_CACHE(oper, n);
+    Obj * cache = 1 + ADDR_OBJ(cacheBag);
+    memmove((void *)(cache + n + 2), (void *)cache,
+            sizeof(Obj) * (CACHE_SIZE - 1) * (n + 2));
+    cache[0] = method;
+    cache[1] = prec;
+    for (UInt i = 0; i < n; i++)
+        cache[2 + i] = ids[i];
+    CHANGED_BAG(cacheBag);
+}
+
+/* These will contain the GAP method selection functions */
+static Obj MethodSelectors[2][7];
+static Obj VerboseMethodSelectors[2][7];
+
+static inline Obj __attribute__((always_inline))
```

Comment:
I thought I had added that in a comment already, but apparently did not :-/ but I don't really like using a raw `__attribute__` like this -- those should always be wrapped in a macro, and covered by an autoconf check. I just did that work in PR #1779, so once that is merged, you may want to rebase this PR, and start using `ALWAYS_INLINE` instead of `__attribute__((always_inline))`

Reply:
Thanks, wil do.

Reply:
PR #1779 has been merged.